core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -81,13 +81,18 @@ private ParallelIterator(
 
     @Override
     public void close() {
+      // close first, avoid new task submit
+      this.closed = true;
+
       // cancel background tasks
       for (int i = 0; i < taskFutures.length; i += 1) {
-        if (taskFutures[i] != null && !taskFutures[i].isDone()) {
-          taskFutures[i].cancel(true);
+        Future<?> taskFuture = taskFutures[i];
+        if (taskFuture != null && !taskFuture.isDone()) {
+          taskFuture.cancel(true);
         }
       }
-      this.closed = true;
+      // clean queue
+      this.queue.clear();
```

Comment:
@rdblue do you remember maybe why the queue wasn't cleared originally when close was called?

Reply:
I think the reason is that we didn't expect the iterator to last very long before getting garbage collected. It seems reasonable to clean up as early as possible though.